Mathlib/RingTheory/AdicCompletion/Exactness.lean

```diff
@@ -0,0 +1,205 @@
+subtracting this delta. -/
+private noncomputable def mapPreimageDelta (x : AdicCauchySequence I N)
+    {n : ℕ} {y yₙ : M} (hy : f y = x (n + 1)) (hyₙ : f yₙ = x n) :
+    { d : (I ^ n • ⊤ : Submodule R M) | f d = f (yₙ - y) } :=
+  have h : f (yₙ - y) ∈ Submodule.map f (I ^ n • ⊤ : Submodule R M) := by
+    rw [Submodule.map_smul'', Submodule.map_top, LinearMap.range_eq_top.mpr hf, map_sub,
+      hyₙ, hy, ← Submodule.neg_mem_iff, neg_sub, ← SModEq.sub_mem]
+    exact AdicCauchySequence.mk_eq_mk (Nat.le_succ n) x
+  ⟨⟨h.choose, h.choose_spec.1⟩, h.choose_spec.2⟩
+
+/- Inductively construct preimage of cauchy sequence. -/
+private noncomputable def mapPreimage (x : AdicCauchySequence I N) :
+    (n : ℕ) → f ⁻¹' {x n}
+  | .zero => ⟨(hf (x 0)).choose, (hf (x 0)).choose_spec⟩
+  | .succ n =>
+      let y := (hf (x (n + 1))).choose
+      have hy := (hf (x (n + 1))).choose_spec
+      let ⟨yₙ, (hyₙ : f yₙ = x n)⟩ := mapPreimage x n
+      let ⟨⟨d, _⟩, (p : f d = f (yₙ - y))⟩ := mapPreimageDelta hf x hy hyₙ
+      ⟨yₙ - d, by simpa [p]⟩
+
+variable (I)
+
+/-- Adic completion preserves surjectivity -/
+theorem map_surjective : Function.Surjective (map I f) := fun y ↦ by
+  apply AdicCompletion.induction_on I N y (fun b ↦ ?_)
+  let a := mapPreimage hf b
+  refine ⟨AdicCompletion.mk I M (AdicCauchySequence.mk I M (fun n ↦ (a n : M)) ?_), ?_⟩
+  · refine fun n ↦ SModEq.symm ?_
+    simp only [SModEq.symm, SModEq, mapPreimage, Submodule.Quotient.mk_sub,
+      sub_eq_self, Submodule.Quotient.mk_eq_zero, SetLike.coe_mem, a]
+  · exact _root_.AdicCompletion.ext fun n ↦ congrArg _ ((a n).property)
+
+end Surjectivity
+
+variable {M : Type u} [AddCommGroup M] [Module R M]
+variable {N : Type u} [AddCommGroup N] [Module R N]
+variable {P : Type u} [AddCommGroup P] [Module R P]
+
+section Injectivity
+
+variable [IsNoetherianRing R] [Module.Finite R N]
+
+open LinearMap
+
+variable (I)
+
+/-- Adic completion preserves injectivity of finite modules over a Noetherian ring. -/
+theorem map_injective {f : M →ₗ[R] N} (hf : Function.Injective f) :
+    Function.Injective (map I f) := by
+  obtain ⟨k, hk⟩ := Ideal.exists_pow_inf_eq_pow_smul I (range f)
+  rw [← LinearMap.ker_eq_bot, LinearMap.ker_eq_bot']
+  intro x
+  apply AdicCompletion.induction_on I M x (fun a ↦ ?_)
+  intro hx
+  refine AdicCompletion.mk_zero_of _ _ _ ⟨42, fun n _ ↦ ⟨n + k, by omega, n, by omega, ?_⟩⟩
+  rw [← Submodule.comap_map_eq_of_injective hf (I ^ n • ⊤ : Submodule R M),
+    Submodule.map_smul'', Submodule.map_top]
+  apply (smul_mono_right _ inf_le_right : I ^ n • (I ^ k • ⊤ ⊓ (range f)) ≤ _)
+  nth_rw 2 [show n = n + k - k by omega]
+  rw [← hk (n + k) (show n + k ≥ k by omega)]
+  refine ⟨?_, ⟨a (n + k), rfl⟩⟩
+  simpa using congrArg (fun x ↦ x.val (n + k)) hx
+
+end Injectivity
+
+section
+
+variable [IsNoetherianRing R] [Module.Finite R N]
+
+variable {f : M →ₗ[R] N} {g : N →ₗ[R] P} (hf : Function.Injective f)
+  (hfg : Function.Exact f g) (hg : Function.Surjective g)
+
+section
```

Comment:
I don't see the point in opening another section here.

Reply:
This is to make the infoview in the proof of `map_exact` less cluttered. The variables introduced in the section with the two aux definitions are local variables that appear in the proof of `map_exact`. I'd prefer keeping the section.

Reply:
OK